core/src/main/java/org/mule/runtime/core/internal/connection/DefaultConnectivityTesterFactory.java

```diff
@@ -0,0 +1,178 @@
+import org.mule.runtime.core.api.retry.policy.RetryPolicyTemplate;
+import org.mule.runtime.core.api.util.func.CheckedRunnable;
+import org.mule.runtime.core.internal.retry.ReconnectionConfig;
+import org.mule.runtime.extension.api.connectivity.NoConnectivityTest;
+import org.mule.runtime.extension.api.runtime.config.ConfigurationInstance;
+
+import java.util.concurrent.locks.Lock;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+
+/**
+ * Implementation of {@link ConnectivityTesterFactory} that takes into account the reconnection strategy.
+ *
+ * @since 4.4
+ */
+public class DefaultConnectivityTesterFactory implements ConnectivityTesterFactory {
+
+  public static final String DO_TEST_CONNECTIVITY_PROPERTY_NAME = "doTestConnectivity";
+
+  private static final Integer DEFAULT_ASYNC_TEST_CONNECTIVITY_TIMEOUT = 30000;
+
+  private static final int ASYNC_TEST_CONNECTIVITY_TIMEOUT =
+      getInteger(ASYNC_TEST_CONNECTIVITY_TIMEOUT_PROPERTY, DEFAULT_ASYNC_TEST_CONNECTIVITY_TIMEOUT);
+
+  private static final Logger LOGGER = getLogger(DefaultConnectivityTesterFactory.class);
+
+  @Inject
+  private ConnectionManagerAdapter connectionManager;
+
+  @Inject
+  private LockFactory lockFactory;
+
+  @Inject
+  private SchedulerService schedulerService;
+
+  private final LazyValue<Boolean> doTestConnectivity = new LazyValue<>(this::getDoTestConnectivityProperty);
+
+  @Override
+  public ConnectivityTester create(String name) {
+    final Lock testConnectivityLock = lockFactory.createLock(this.getClass().getName() + "-testConnectivity-" + name);
+    return new ConnectivityTester() {
+
+      @Override
+      public void testConnectivity(ConnectionProvider provider, ConfigurationInstance configurationInstance)
+          throws MuleException {
+        if (provider instanceof NoConnectivityTest || !doTestConnectivity.get()) {
+          return;
+        }
+
+        Scheduler retryScheduler = schedulerService.ioScheduler();
+        RetryPolicyTemplate retryTemplate = connectionManager.getRetryTemplateFor(provider);
+        ReconnectionConfig reconnectionConfig = connectionManager.getReconnectionConfigFor(provider);
+        final Latch latch = new Latch();
+        RetryCallback retryCallback = new RetryCallback() {
+
+          @Override
+          public void doWork(RetryContext context) throws Exception {
+            try {
+              Lock lock = testConnectivityLock;
+              if (lock != null) {
+                final boolean lockAcquired = lock.tryLock();
+                if (lockAcquired) {
+                  LOGGER.debug("Doing testConnectivity() for config '{}'", name);
+                  try {
+                    ConnectionValidationResult result = connectionManager.testConnectivity(configurationInstance);
+                    if (result.isValid()) {
+                      context.setOk();
+                    } else {
+                      if ((reconnectionConfig.isFailsDeployment())) {
```

Comment:
if this does not affect performance (I don't think so), I would extract the case where the test connectivity fails at least to a method to make the general doWork method more declarative and avoid problems if this logics gets more troublesome.